File: app/src/main/java/com/innoveworkshop/gametest/MainActivity.java
package com.innoveworkshop.gametest;

import androidx.appcompat.app.AppCompatActivity;

import android.graphics.Color;
import android.os.Bundle;
import android.view.MotionEvent;
import android.view.View;
import android.widget.Button;

import com.innoveworkshop.gametest.assets.DroppingRectangle;
import com.innoveworkshop.gametest.engine.Circle;
import com.innoveworkshop.gametest.engine.GameObject;
import com.innoveworkshop.gametest.engine.GameSurface;
import com.innoveworkshop.gametest.engine.Rectangle;
import com.innoveworkshop.gametest.engine.Vector;

public class MainActivity extends AppCompatActivity {
    protected GameSurface gameSurface;
    protected Button upButton;
    protected Button downButton;
    protected Button leftButton;
    protected Button rightButton;

    protected Game game;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        gameSurface = (GameSurface) findViewById(R.id.gameSurface);
        game = new Game();
        gameSurface.setRootGameObject(game);

        setupControls();
    }

    private void setupControls() {
        upButton = (Button) findViewById(R.id.up_button);
        upButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                game.circle.position.y -= 10;
            }
        });

        downButton = (Button) findViewById(R.id.down_button);
        downButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                game.circle.position.y += 10;
            }
        });

        leftButton = (Button) findViewById(R.id.left_button);
        leftButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                game.circle.position.x -= 10;
            }
        });

        rightButton = (Button) findViewById(R.id.right_button);
        rightButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                game.circle.position.x += 10;
            }
        });
    }

    class Game extends GameObject {
        public Circle circle;

        @Override
        public void onStart(GameSurface surface) {
            super.onStart(surface);

            circle = new Circle(surface.getWidth() / 2, surface.getHeight() / 2, 100, Color.RED);
            surface.addGameObject(circle);

            surface.addGameObject(new Rectangle(new Vector(surface.getWidth() / 3, surface.getHeight() / 3),
                    200, 100, Color.GREEN));

            surface.addGameObject(new DroppingRectangle(new Vector(surface.getWidth() / 3, surface.getHeight() / 3),
                    100, 100, 10, Color.rgb(128, 14, 80)));
        }

        @Override
        public void onFixedUpdate() {
            super.onFixedUpdate();

            if (!circle.isFloored() && !circle.hitRightWall())
                circle.setPosition(circle.position.x + 1, circle.position.y + 1);
        }
    }
}

File: app/src/main/java/com/innoveworkshop/gametest/MainActivity.kt
package com.innoveworkshop.gametest

import android.graphics.Color
import android.graphics.PointF
import android.os.Bundle
import android.view.View
import android.widget.Button
import androidx.appcompat.app.AppCompatActivity
import com.innoveworkshop.gametest.assets.Bin
import com.innoveworkshop.gametest.engine.Circle
import com.innoveworkshop.gametest.engine.GameObject
import com.innoveworkshop.gametest.engine.GameSurface
import com.innoveworkshop.gametest.engine.Rigidbody

import com.innoveworkshop.gametest.engine.Vector

class MainActivity : AppCompatActivity() {
    protected var gameSurface: GameSurface? = null
    protected var upButton: Button? = null
    var rigidbody: Rigidbody? = null
    protected var game: Game? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        gameSurface = findViewById<View>(R.id.gameSurface) as GameSurface
        game = Game()
        gameSurface!!.setRootGameObject(game)

        setupControls()
    }

    private fun setupControls() {
        upButton = findViewById<View>(R.id.up_button) as Button
        upButton!!.setOnClickListener { game!!.circle!!.rigidbody.applyForce(PointF(0f, -700f)) }

    }

    inner class Game : GameObject() {
        var circle: Circle? = null
        var bin: Bin? = null
        override fun onStart(surface: GameSurface?) {
            super.onStart(surface)

            //val bin = Bin(100f, 100f, 20f, 20f)
            //surface!!.addGameObject(bin)
            circle = Circle(
                (surface!!.width / 2).toFloat(),
                (surface.height / 2).toFloat(),
                30f,
                Color.RED
            )

            surface.addGameObject(circle!!)

             //Initialize the bin
            val binX = (surface.width / 5).toFloat() // Centered horizontally
            val binY = (surface.height / 4).toFloat()           // Positioned at the bottom
            bin = Bin(binX, binY, 300f, 200f)

        }


        override fun onFixedUpdate() {
            super.onFixedUpdate()

            val deltaTime = 7.5f / 30f
//            circle!!.rigidbody.applyForce(PointF(0f, 9.8f))
//            circle!!.updatePhysics(deltaTime)
//
//
//            circle!!.position.x += circle!!.rigidbody.velocity.x * deltaTime
//            circle!!.position.y += circle!!.rigidbody.velocity.y * deltaTime
//
//            if (circle!!.hitLeftWall() || circle!!.hitRightWall()) {
//                circle!!.rigidbody.velocity.x = -circle!!.rigidbody.velocity.x
//            }
//            if (circle!!.hitCeiling() || circle!!.isFloored) {
//                circle!!.rigidbody.velocity.y = -circle!!.rigidbody.velocity.y
//            }
//

        }
    }
}